File: Irrgarten_Ruby/Labyrinth.rb
#encoding:utf-8

require_relative 'Dice'

module Irrgarten 
	class Labyrinth
		@@BLOCK_CHAR ='X'
		@@EMPTY_CHAR = '-'
		@@MONSTER_CHAR = 'M'
		@@COMBAT_CHAR = 'C'
		@@EXIT_CHAR = 'E'
		@@ROW = 0
		@@COL = 1
		
		attr_accessor :labyrinth
		
		private
	
		def pos_ok ( row, col) #int, int
			ok=false
			if (row >=0 && col>=0 && row < @n_rows && col< @n_cols) 
				ok=true
			end
			return ok	
		end
	
		def empty_pos ( row, col) #int, int
			empty=false
			if (pos_ok(row,col)) 
				if( @labyrinth[row][col]==@@EMPTY_CHAR) 
					empty=true
				end
			end
			return empty
		end
	
		def monster_pos (row, col) #int, int
			monster=false
			if (pos_ok(row,col)) 
				if( @labyrinth[row][col]==@@MONSTER_CHAR) 
					monster=true
				end
			end
			return monster
		end
		
		def exit_pos (row, col) #int, int
			exit=false
			if (pos_ok(row,col)) 
				if( @labyrinth[row][col]==@@EXIT_CHAR) 
					exit=true
				end
			end
			return exit
		end
		
		def combat_pos (row, col) #int, int
			combat=false
			if (pos_ok(row,col)) 
				if( @labyrinth[row][col]==@@COMBAT_CHAR) 
					combat=true
				end
			end
			return combat
		end
		
		def can_step_on (row, col) #int, int
			can=false
			if (pos_ok(row,col) && (empty_pos(row,col)||monster_pos(row,col) || exit_pos(row,col))) 				 
				can=true
			end
			return can
		end
		
		def update_old_pos(row, col) #int,int
			if( pos_ok(row,col) )
				if (@labyrinth[row][col]==@@COMBAT_CHAR) 
					@labyrinth[row][col]=@@MONSTER_CHAR
				end
			else 
				@labyrinth[row][col]=@@EMPTY_CHAR
			end
		end
		
		def dir_2_pos(row,col,direction)#int,int, Directions		
			fila=row
			columna=col
			case direction
				when Directions::UP
					fila-=1
				when Directions::DOWN
					fila+=1
				when Directions::RIGHT
					columna+=1
				when Directions::LEFT
					columna-=1
			end
			return [fila,columna]
		end
		
		def random_empty_pos()
			fila=Dice.random_pos(@n_rows)
			columna=Dice.random_pos(@n_cols)
			while (@labyrinth[fila][columna]!=@@EMPTY_CHAR)
				fila=Dice.random_pos(@n_rows)
				columna=Dice.random_pos(@n_cols)
			end
			return [fila,columna]
		end
		
		def put_player_2d (old_row,old_col,row,col,player)#int,int,int,int,Player
			output = nil
			
			if (can_step_on(row,col))
				if (pos_ok(old_row,old_col))
					if (@players[old_row][old_col]==player)
						update_old_pos(old_row,old_col)
						@players[old_row][old_col]=nil
					end
				end
				
				if(monster_pos(row,col))
					@labyrinth[row][col]=@@COMBAT_CHAR
					output=@monsters[row][col]
				else
					number = player.number
					@labyrinth[row][col]=number
				end
				
				@players[row][col]=player
				player.set_pos(row,col)
			end
			output
		end
		
		
		
		public
		
		def initialize( n_rows,n_cols,exit_row,exit_col) #int,int,int,int
			@n_rows=n_rows
			@n_cols=n_cols
			@exit_row=exit_row
			@exit_col=exit_col
			
			@labyrinth = Array.new(n_rows){Array.new(n_cols){@@EMPTY_CHAR}}
			
			@monsters = Array.new(n_rows){Array.new(n_cols){nil}}
			
			@players = Array.new(n_rows){Array.new(n_cols){nil}}
			
			@labyrinth[exit_row][exit_col]=@@EXIT_CHAR
		end
		
		def spread_players(players)
  			players.each do |p|
    				pos = random_empty_pos
    				put_player_2d(-1, -1, pos[@@ROW], pos[@@COL], p)
  			end
		end
		
		def have_a_winner()
			ganador=false
			if( @players[@exit_row][@exit_col] != nil) 
				ganador=true
			end
			return ganador
		end
		
		def to_s()
			salida = "Nº de filas: #{@n_rows}, Nº de columnas: #{@n_cols} \n"
			for fila in @labyrinth
				for i in fila
					salida+= i
				end 
				salida += "\n"
			end
			salida
		end
		
		def add_monster ( row, col, monster)#int,int,monster
			if((pos_ok(row,col)) && (empty_pos(row,col))) 
				@monsters[row][col] = monster
				@labyrinth[row][col]=@@MONSTER_CHAR
			end
		end
		
		def put_player (direction, player) #Direction, Player
			old_row = player.row
			old_col = player.col
			new_pos = dir_2_pos(old_row,old_col,direction)
			monster = put_player_2d(old_row,old_col,new_pos[@@ROW],new_pos[@@COL],player)
			monster
		end
		
		def add_block (orientation, start_row, start_col, length) #Orientation, int ,int ,int
			inc_row = 0
			inc_col = 0
			if (orientation == Orientation::VERTICAL)
				inc_row = 1
			else
				inc_col = 1
			end
			
			col = start_col
			row = start_row
			
			while (pos_ok(row,col) && empty_pos(row,col) && length > 0)
				@labyrinth[row][col]=@@BLOCK_CHAR
				length -= 1
				row += inc_row
				col += inc_col
			end
		end
		
		def valid_moves (row,col) #int,int
			output = Array.new
			
			if (can_step_on(row+1,col))
				output.push(Directions::DOWN)
			end
			
			if (can_step_on(row-1,col))
				output.push(Directions::UP)
			end
			
			if (can_step_on(row,col+1))
				output.push(Directions::RIGHT)
			end
			
			if (can_step_on(row,col-1))
				output.push(Directions::LEFT)
			end
			
			output
		end
	end
end

File: Irrgarten_Java/Irrgarten_P4/src/irrgarten/Labyrinth.java

package irrgarten;

import java.util.ArrayList;

public class Labyrinth {
    
    private static char BLOCK_CHAR = 'X';
    private static char EMPTY_CHAR = '-';
    private static char MONSTER_CHAR = 'M';
    private static char COMBAT_CHAR = 'C';
    private static char EXIT_CHAR = 'E';
    private static int ROW = 0;
    private static int COL = 1;
    
    private int nRows;
    private int nCols;
    private int exitRow; //Fila en la que está la salida
    private int exitCol; //Columna en la que está la salida
    
    private Monster[][] monsters; //Array que representa las posiciones del laberinto en las que hay monstruos
    private Player[][] players; //Array que representa las posiciones del laberinto en las que hay jugadores
    private char[][] labyrinth; //Array que representa las posiciones del laberinto
    
    private boolean posOK(int row, int col){
        return (row >= 0 && col >= 0 && row < nRows && col < nCols); 
    }
    
    private boolean emptyPos (int row, int col){
        if (posOK(row,col)){
            return labyrinth[row][col]==Labyrinth.EMPTY_CHAR;
        }
        else 
            return false;
    }
    
    private boolean monsterPos (int row, int col){
        if (posOK(row,col)){
            return labyrinth[row][col]==Labyrinth.MONSTER_CHAR;
        } else 
            return false;
    }
    
    private boolean exitPos (int row, int col){
        if (posOK(row,col)){
            return labyrinth[row][col]==Labyrinth.EXIT_CHAR;
        }
        else 
            return false;
    }
    
    private boolean combatPos(int row, int col){
        if (posOK(row,col)){
            return labyrinth[row][col]==Labyrinth.COMBAT_CHAR;
        }
        else 
            return false;    
    }
    
    private boolean canStepOn (int row, int col){
        if (posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col))){
            return true;
        } else
            return false;
    }
    
    private void updateOldPos (int row, int col){
        if (posOK(row,col)){
            if (labyrinth[row][col]==Labyrinth.COMBAT_CHAR){
                labyrinth[row][col]=Labyrinth.MONSTER_CHAR;
            } else {
                labyrinth[row][col]=Labyrinth.EMPTY_CHAR;
            } 
        }
    }
    
    private int[] dir2Pos(int row, int col, Directions direction){
        int fila=row, columna=col;
        switch (direction){
            case UP -> fila--;
            case DOWN -> fila++;
            case RIGHT -> columna++;
            case LEFT -> columna--;
        }
        return new int[]{fila,columna};
    }
    
    private int[] randomEmptyPos(){
        int fila, columna;
        fila = Dice.randomPos(nRows);
        columna = Dice.randomPos(nCols);
        while (labyrinth[fila][columna]!=Labyrinth.EMPTY_CHAR) {
            fila = Dice.randomPos(nRows);
            columna = Dice.randomPos(nCols);
        }
        return new int[]{fila,columna};
    }
    
    private Monster putPlayer2D (int oldRow, int oldCol, int row, int col, Player player){
        Monster output=null;
        
        if (canStepOn(row,col)){
            if(posOK(oldRow,oldCol)){
                if(this.players[oldRow][oldCol]==player){
                    updateOldPos(oldRow,oldCol);
                    this.players[oldRow][oldCol]=null; 
                }
            }
        
            if(monsterPos(row,col)){
                this.labyrinth[row][col]=Labyrinth.COMBAT_CHAR;
                output=monsters[row][col];  
            }
            else{
                char number=player.getNumber();
                this.labyrinth[row][col]=number;
            }
            
            this.players[row][col]=player;
            player.setPos(row, col);

        }
        return output;

    }
    
    private void set(int row, int col, char block){ //NECESARIO?
            this.labyrinth[row][col]=Labyrinth.BLOCK_CHAR; 
    }
    
    public Labyrinth (int nRows, int nCols, int exitRow, int exitCol){
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        
        this.monsters = new Monster[nRows][nCols];
        this.players = new Player[nRows][nCols];
        this.labyrinth = new char[nRows][nCols];
        
        for (int i=0; i<nRows;i++){
            for (int j=0; j<nCols; j++){
                monsters[i][j] = null;
                players[i][j] = null;
                labyrinth[i][j] = Labyrinth.EMPTY_CHAR;
            }
        }
        
        labyrinth[exitRow][exitCol] = Labyrinth.EXIT_CHAR;
    }
    
    public void spreadPlayers(ArrayList<Player> players){
        for (int i=0; i<players.size(); i++){
            Player p= players.get(i);
            
            int[] pos = randomEmptyPos();
            putPlayer2D(-1,-1,pos[ROW],pos[COL],p);
        }
    }
    
    public boolean haveAWinner(){
        if (players[exitRow][exitCol] != null){
            return true;
        } else
            return false;
    }
    
    @Override
    public String toString(){
        String salida = "Nº de filas: " + nRows + " Nº Columnas: " + nCols + " Casilla de meta: (" + exitRow + "," + exitCol + ")" + "\n Laberinto: \n";
        for (int i=0; i<nRows;i++){
            for (int j=0; j<nCols; j++){
                salida += labyrinth[i][j] + " ";
            }
            salida += "\n";
        }
        return salida;
    }
    
    public void addMonster(int row, int col, Monster monster){
        if (posOK(row,col) && emptyPos(row,col)){
            monsters[row][col]=monster;
            labyrinth[row][col]=Labyrinth.MONSTER_CHAR;
        }
    }
    
    public Monster putPlayer(Directions direction, Player player){ //Que significa este error?
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        int[] newPos = dir2Pos(oldRow, oldCol, direction);
        Monster monster = putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player);
        
        return monster;
    }
    
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow=0, incCol=0;
        if(orientation==Orientation.VERTICAL){
            incRow=1;
        } else {
            incCol=1;
        }
        
        int col=startCol, row=startRow;
        
        while(posOK(row,col) && emptyPos(row,col) && length>0){
            set(row,col,BLOCK_CHAR);
            length--;
            row+=incRow;
            col+=incCol;
        }
            
    }
    
    public ArrayList<Directions> validMoves (int row, int col){
        ArrayList<Directions> output = new ArrayList<>();
        
        if (canStepOn(row+1,col)){
            output.add(Directions.DOWN);
        }
        if (canStepOn(row-1,col)){
            output.add(Directions.UP);
        }
        if (canStepOn(row,col+1)){
            output.add(Directions.RIGHT);
        }
        if (canStepOn(row,col-1)){
            output.add(Directions.LEFT);
        }
        return output;
    }
    
    public int getNRows(){
        return nRows;
    }
    
    public int getNCols(){
        return nCols;
    }
    
}
